lib/services/publish-service.ts

```diff
@@ -0,0 +1,211 @@
+﻿///<reference path="../.d.ts"/>
+"use strict";
+
+import options = require("../common/options");
+import os = require("os");
+import path = require("path");
+import util = require("util");
+
+export class PublishService implements IPublishService {
+    private static JSON_PUBLISH_FILE_NAME = ".abpublish";
+    private static ALLOWED_COONECTION_TYPE = "ftp";
+    private static ERROR_MESSAGE_WHEN_PROJECT_EXISTS_ON_REMOTE = "User cancelled operation";
+    private static PUBLIC_URL_DEFAULT_VALUE = "";
+
+    public allPublishConnections: IPublishConnection[];
+    public ftpPublishConnections: IPublishConnection[];
+
+    constructor(private $errors: IErrors,
+                private $fs: IFileSystem,
+                private $logger: ILogger,
+                private $progressIndicator: IProgressIndicator,
+                private $projectConstants: Project.IProjectConstants,
+                private $projectFilesManager: Project.IProjectFilesManager,
+                private $project: Project.IProject,
+                private $prompter: IPrompter,
+                private $server: Server.IServer) {
+
+        if (!this.$project.capabilities.publish) {
+            this.$errors.failWithoutHelp("The publish command is only applicable to projects of type %s", this.$projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.MobileWebsite);
+        }
+
+        this.readConnections().wait();
+        this.ftpPublishConnections = _.filter(this.allPublishConnections, c => c.type === PublishService.ALLOWED_COONECTION_TYPE);
+    }
+
+    public publish(idOrUrl: string, username: string, password: string): IFuture<void> {
+        return (() => {
+            var projectName = this.$project.projectData.ProjectName;
+            var ftpConnectionData: Server.OfflineClientFtpConnectionData = {
+                RemoteUrl: idOrUrl,
+                ShouldPurge: options.force,
+                Username: username,
+                Password: password
+            };
+
+            if (!ftpConnectionData.Username) {
+                ftpConnectionData.Username = this.$prompter.getString("Username:").wait();
+            }
+
+            if (!ftpConnectionData.Password) {
+                ftpConnectionData.Password = this.$prompter.getPassword("Password:", {allowEmpty: true}).wait();
+            }
+
+            var id = parseInt(idOrUrl);
+
+            // the check wether it contains a dot is performed
+            // in case the user tries to publish to an IP address
+            if (!isNaN(id) && idOrUrl.indexOf('.') < 0) {
+                this.validateId(id);
+
+                // (id - 1) is the index of the object to be deleted
```

Comment:
the comment says that the connection will be deleted, but the method is called publish. What are we going to do with this connection? :smirk_cat: